Increase robustness of mods priority in profile.

Holt59 opened the pull request:
This is a follow up from #1373, the goal is to clean the priority management in the profile.

## Major

### Profile

The major changes are in `refreshModStatus`, where I added a big comment to explain the priority management and try to clean the code and make it a bit more readable. The major changes are

- `m_NumRegularMods` is computed in `refreshModStatus`, correctly corresponds to what the doc says (number of mods with modifiable priority), and is not exposed outside of the profile.
- The Overwrite mod now has a priority in the profile, set to the highest priority (= number of mods - 1). Before it was (likely, not 100% sure) the only mod with -1. In particular the comment in `updateIndices` was not accurate as the Overwrite was indeed in `m_ModIndexByPriority` (that's how the by-priority proxy retrieves it).
- Backup mods are manually put at the end of the priority. Before, backups could ends up in between mods. That was particularly noticeable when you would install a mod while having backups. If the name of the mod was after some backup names (alphabetically), then you'd end up with a gap at the end of the priority list.

Another change is that `Profile::setModPriority` now clamps the priority properly, so you don't need to check beforehand if the priority is valid. You cannot set priority of backups (it was possible before, although you probably could only do it from a plugin).

### Directory Structure

- Before, the priority in the directory structure was either contiguous (for `refresh()`) or not (when changing priority of a mod). Now the priorities in the directory structure are always consistent with the one in the profile (with a +1 offset for `<data>`).
- Since overwrite has a proper priority in the profile now, it prevent some issues/bugs. Currently (before this PR), after changing the priority of a mod, the overwrite gets a priority of 0 in the directory structure (this is hardly visible since the data-tab is not refreshed and if you click refresh then priority are re-computed).

### Mods & Mod List

- I removed `getFixedPriority` because it was misleading (for me) as backups have a fixed priority of `-1` but ends up with different priorities in the profile... I added `hasAutomaticPriority` which returns `true` for backups and overwrites and means that the priority is not user-modifiable. 
- I added `isAlwaysDisabled`, similar to `isAlwaysEnabled` to avoid some issues (see minor fixes).
- The `ModList` now directly expose the priority from the profile, nothing custom like before where backups would have -1 priority and overwrite `INT_MAX`. The "sorting" is done in the sort proxy to put overwrite after all mods and backups before all mods.

## Minor

- The "Enable all" action does not enable backups anymore (see `isAlwaysDisabled`).
- The "Enable/disable all" action now displays the actual number of mods that are going to be enabled/disabled.
- The data tab is refreshed automatically when the priority of a mod changes or when a mod is enabled/disabled.

isanae commented on src/modlist.cpp:
```diff
@@ -223,8 +197,7 @@ QVariant ModList::data(const QModelIndex &modelIndex, int role) const
     else if (column == COL_PRIORITY) {
-      int priority = modInfo->getFixedPriority();
-      if (priority != INT_MIN) {
+      if (modInfo->isBackup() || modInfo->isOverwrite()) {
```
Why not `isFixedPriority()`?

Holt59 replied:
Because I added `isFixedPriority` after and forgot this part, will fix.

isanae commented on src/modlist.cpp:
```diff
@@ -337,9 +310,11 @@ QVariant ModList::data(const QModelIndex &modelIndex, int role) const
-      int priority = modInfo->getFixedPriority();
-      if (priority != INT_MIN) {
-        return priority;
+      if (modInfo->isBackup()) {
```
Can these calls be put encapsulated somewhere? You're duplicating this code in `COL_PRIORITY` and `PriorityRole` below. Maybe `Profile::getModPriority()` should do that instead? Or add something like `ModInfo::getPriority(Profile&)`? I'm thinking if we ever add another type of mod with fixed prio.

Holt59 replied:
The "problem" is that these are completely modlist specific. The profile does not use and care about these "custom" priorities for backups and overwrite so putting something in the profile would be very weird. I'll add a `ModList` member function.

isanae commented on src/profile.cpp:
```diff
@@ -270,10 +269,9 @@ void Profile::createTweakedIniFile()
-    unsigned int idx = modIndexByPriority(i);
-    if (m_ModStatus[idx].m_Enabled) {
-      ModInfo::Ptr modInfo = ModInfo::getByIndex(idx);
+  for (auto& [priority, index] : m_ModIndexByPriority) {
```
This should probably be `const auto&`.

isanae commented on src/modlist.cpp:
```diff
@@ -1074,14 +1047,14 @@ int ModList::dropPriority(int row, const QModelIndex& parent) const
-    if ((row < 0) || (row > static_cast<int>(m_Profile->numRegularMods()))) {
-      newPriority = m_Profile->numRegularMods() + 1;
+    if (row < 0 || row >= rowCount()) {
+      newPriority = std::numeric_limits<int>::max();
```
One thing that's getting pretty confusing is the usage of `INT_MIN`, `UINT_MAX`, `std::numeric_limits<int>::max()`, etc. This line I think means "at the end"? I know we're already using these constants all over the place, but I hate it.

Holt59 replied:
I removed usage of `INT_MIN` to mean "not a fixed priority" because I agree it was misleading. Here the value means "the biggest you can", not really "at the end" (well depends what you call "at the end"), and it let the profile fix it.

isanae commented:
Are you sure you changed everything that assumed the overwrite prio was some sort of sentinel value? Stuff like [this](https://github.com/ModOrganizer2/modorganizer/blob/7c9d3777c28f79b1ae7cdbda107493493ca1ba2a/src/profile.cpp#L594)?

Holt59 commented:
> 
> 
> Are you sure you changed everything that assumed the overwrite prio was some sort of sentinel value? Stuff like [this](https://github.com/ModOrganizer2/modorganizer/blob/7c9d3777c28f79b1ae7cdbda107493493ca1ba2a/src/profile.cpp#L594)?

I will do a thorough check. The part you linked does not require change (maybe replacing `INT_MAX` with `numeric_limits`).

isanae commented:
Just checking. I still have nightmares of trying to work on this stuff and having Data sometimes being a mod with prio 0 and sometimes not, and overwrite being `INT_MAX`. That was in my branch for refactoring the register stuff, I abandoned it because it introduced a bunch of subtle bugs that I couldn't fix.

Holt59 commented:
> 
> 
> Are you sure you changed everything that assumed the overwrite prio was some sort of sentinel value? Stuff like [this](https://github.com/ModOrganizer2/modorganizer/blob/7c9d3777c28f79b1ae7cdbda107493493ca1ba2a/src/profile.cpp#L594)?

Also there are two "sentinel" values for Overwrite. 
- In the Profile, the priority was `-1` and is now `ModInfo::getNumMods() - 1`. I changed it to be consistent but I can switch back to -1 without breaking any changes.
- In the Modlist, and almost everywhere where the priority of overwrite is exposed, the value is `INT_MAX`, and I did not change that one (except switching to `numeric_limits<int>::max()` in some places).

Holt59 commented:
> 
> 
> Just checking. I still have nightmares of trying to work on this stuff and having Data sometimes being a mod with prio 0 and sometimes not, and overwrite being `INT_MAX`. That was in my branch for refactoring the register stuff, I abandoned it because it introduced a bunch of subtle bugs that I couldn't fix.

I think I'm done, I tried to clean a lot of stuffs. I looked at (I think) all usages of the priority, from either the profile or the mod list, and tried to fix everything.

- I removed the priority for `GroupingRole` in the mod list because it's not used anywhere.
- For the priority role, I always return the profile priority, and I let the sort proxy do the actual sorting between backups, overwrite, etc.
- I removed a bunch of unused stuffs (functions, variables).

Having overwrite with a proper priority in the profile fixes a bug where after a priority change (drag&drop, send to, etc.), the overwrite would get a priority of 0 in the directory structure.

I noticed that in the directory structure, the priority from the mod list is not actually used, but contiguous priority are computed, while in `ModListView::onModPrioritiesChanged`, priorities from the mod list are used (not contiguous, and this is where the overwrite bug comes from). I kept it that way, although it is probably now possible to use the profile priority in the directory structure refresh.

isanae commented on src/mainwindow.cpp:
```diff
@@ -313,6 +313,12 @@ MainWindow::MainWindow(Settings &settings
 
   setupModList();
   ui->espList->setup(m_OrganizerCore, this, ui);
+  connect(m_OrganizerCore.modList(), &ModList::modPrioritiesChanged, [this]() {
```
I don't think we typically update the Data tab automatically, do we? There should be plenty of other places where we don't update it, so I'm not sure this is worth it.

Holt59 replied:
Hmm, seems you're right. The datatab is only updated if it's the currently visible tab when the directory structure is refreshed, which does not include mod removal for instance. I'll see if I can make it more consistent or I'll remove it from the PR.

isanae replied:
The data tab used to reset completely when refreshing iirc, and it was slow as heck, which is probably why it wasn't refreshing automatically. With the stuff I did, it might be possible to keep it up to date now. Maybe.

Holt59 replied:
I will time the `updateTree` (if it's not already) and check with some of my mod setup. What I think I'll do (if I manage) is to only refresh when the tab is visible, and otherwise set some kind of flag to refresh when the tab is made visible so that it should have no impact if you have another tab visible.

isanae replied:
There's already stuff in `MainWindow::on_tabWidget_currentChanged()` that calls `DataTab::activated()`, you can probably reuse it.
